### PR TITLE
PoS update snapshot epoch duration

### PR DESCRIPTION
**Pull request by AeonSw4n**

### CODE REVIEW

**Comment by AeonSw4n** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1075?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* **#1077** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1077?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1076** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1076?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1075** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1075?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1074** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1074?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `feature/pos-hypersync-updates`

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/1075?utm_source=stack-comment).

File: lib/snapshot.go [hunk 1]
```diff
@@ -666,7 +666,16 @@ func (snap *Snapshot) FinishProcessBlock(blockNode *BlockNode) {
...
 	defer snap.CurrentEpochSnapshotMetadata.updateMutex.Unlock()
-	if uint64(blockNode.Height)%snap.SnapshotBlockHeightPeriod == 0 &&
+
+	// If the block height is divisible by the snapshot block height period, we update the snapshot metadata.
+	// For PoW blocks, until the first PoS fork height, we use the default snapshot block height period of 1000.
+	// For blocks after the first PoS fork height, we use the snapshot block height period defined in the params.
+	snapshotBlockHeightPeriod := snap.SnapshotBlockHeightPeriod
+	if blockNode.Height < snap.params.ForkHeights.ProofOfStake1StateSetupBlockHeight {
+		snapshotBlockHeightPeriod = snap.params.DefaultPoWSnapshotBlockHeightPeriod
+	}
```
Comment (lazynina) [inline lib/snapshot.go:676]: small request - can we make a function on params called `GetSnapshotBlockHeightPeriod` that does this conditional logic for fetching the PoS vs. PoW SnapshotBlockHeightPeriod? Will be nice in the event we need to use this in more places.

Comment (tholonious) [reply lib/snapshot.go:676]: +1 to this. That way, you can reuse it in server.go line 1034

File: lib/server.go [hunk 2]
```diff
@@ -1028,6 +1028,11 @@ func (srv *Server) _handleHeaderBundle(pp *Peer, msg *MsgDeSoHeaderBundle) {
 			if srv.cmgr.HyperSync {
 				bestHeaderHeight := uint64(srv.blockchain.headerTip().Height)
 				expectedSnapshotHeight := bestHeaderHeight - (bestHeaderHeight % srv.snapshot.SnapshotBlockHeightPeriod)
+				// The peer's snapshot block height period before the first PoS fork height is expected to be the
+				// PoW default value. After the fork height, it's expected to be the value defined in the params.
+				if bestHeaderHeight < uint64(srv.params.ForkHeights.ProofOfStake1StateSetupBlockHeight) {
+					expectedSnapshotHeight = bestHeaderHeight - (bestHeaderHeight % srv.params.DefaultPoWSnapshotBlockHeightPeriod)
+				}
```
Comment (tholonious) [inline lib/server.go:1035]: Reuse the function Nina's comment here so you don't have to make the same block height assumption here

**Review by tholonious — CHANGES_REQUESTED** [review]

**Review by tholonious — APPROVED** [review]

**Pull request merged**